File: 算法/代码随想录算法PDF/00调试源代码/530.二叉搜索树的最小绝对差.py
#
# @lc app=leetcode.cn id=530 lang=python3
#
# [530] 二叉搜索树的最小绝对差
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        minDiff = float('inf')
        pre = None
        st = []
        while root or st:
            if root:
                st.append(root)
                root = root.left
            else:
                root = st.pop()
                if pre:
                    minDiff = min(minDiff, root.val - pre.val)
                pre = root
                root = root.right
        return minDiff
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/530.二叉搜索树的最小绝对差.cpp
/*
 * @lc app=leetcode.cn id=530 lang=cpp
 *
 * [530] 二叉搜索树的最小绝对差
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        TreeNode* pre = nullptr;  // 记录前一个节点
        stack<TreeNode*> st;
        int minDiff = INT_MAX;
        while (root || !st.empty()) {
            if (root) {
                st.push(root);
                root = root->left;
            } else {
                root = st.top(); st.pop();
                if (pre) {
                    minDiff = min(minDiff, root->val - pre->val);
                }
                pre = root;
                root = root->right;
            }
        }
        return minDiff;
    }
};
// @lc code=end

